### PR TITLE
Enable asset helpers in precompiled templates 

### PR DESCRIPTION
**Pull request by yasaichi**
Currently, precompiling Slim templates fails due to `undefined local variable or method` error when asset helpers (e.g. `stylesheet_link_tag`) are used in them.
This PR fixes that by passing `Sprockets::Context` object to `Slim::Template#render`.

### CODE REVIEW

**Comment by januszm** [comment]
@yasaichi could you give us more information about the nature of this problem? I've just created a new Rails 5 app with slim-rails 3.1.1 and `stylesheet_link_tag` works fine in layout

**Comment by yasaichi** [comment]
@januszm 
Sorry for my unclear expression :bow:
To give you more information, I created a demo app to repro the bug:
https://github.com/yasaichi/slim_rails_demo

Could you please execute `bin/rails assets:precompile` on the `repro-pulls-135` branch?
Perhaps you could get an error like the followings:

```
$ bin/rails assets:precompile
I, [2017-02-15T23:32:02.587213 #91310]  INFO -- : Writing /path/to/slim_rails_demo/public/assets/application-4586df2c9172aec6cc707426ba5f83f7a590a4aeac2fe816fa5023c32963f2f9.js
I, [2017-02-15T23:32:02.593717 #91310]  INFO -- : Writing /path/to/slim_rails_demo/public/assets/application-4586df2c9172aec6cc707426ba5f83f7a590a4aeac2fe816fa5023c32963f2f9.js.gz
I, [2017-02-15T23:32:02.644337 #91310]  INFO -- : Writing /path/to/slim_rails_demo/public/assets/application-af04b226fd7202dfc532ce7aedb95a0128277937e90d3b3a3d35e1cce9e16886.css
I, [2017-02-15T23:32:02.644564 #91310]  INFO -- : Writing /path/to/slim_rails_demo/public/assets/application-af04b226fd7202dfc532ce7aedb95a0128277937e90d3b3a3d35e1cce9e16886.css.gz
rails aborted!
NoMethodError: undefined method `stylesheet_link_tag' for #<Object:0x007f9c5fe9b3a0>
test:4:in `block in singleton class'
test:-2:in `instance_eval'
test:-2:in `singleton class'
test:-4:in `__tilt_70154784602600'
/path/to/vendor/bundle/ruby/2.3.0/gems/tilt-2.0.6/lib/tilt/template.rb:161:in `call'
/path/to/vendor/bundle/ruby/2.3.0/gems/tilt-2.0.6/lib/tilt/template.rb:161:in `evaluate'
/path/to/vendor/bundle/ruby/2.3.0/gems/tilt-2.0.6/lib/tilt/template.rb:100:in `render'
/path/to/vendor/bundle/ruby/2.3.0/bundler/gems/slim-rails-389949934b9d/lib/slim-rails/register_engine.rb:6:in `call'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/processor_utils.rb:75:in `call_processor'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/processor_utils.rb:57:in `block in call_processors'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/processor_utils.rb:56:in `reverse_each'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/processor_utils.rb:56:in `call_processors'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/loader.rb:134:in `load_from_unloaded'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/loader.rb:60:in `block in load'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/loader.rb:44:in `load'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/cached_environment.rb:20:in `block in initialize'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/cached_environment.rb:47:in `load'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/base.rb:66:in `find_asset'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/base.rb:73:in `find_all_linked_assets'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/manifest.rb:142:in `block in find'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/legacy.rb:114:in `block (2 levels) in logical_paths'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/path_utils.rb:228:in `block in stat_tree'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/path_utils.rb:212:in `block in stat_directory'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/path_utils.rb:209:in `each'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/path_utils.rb:209:in `stat_directory'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/path_utils.rb:227:in `stat_tree'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/legacy.rb:105:in `each'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/legacy.rb:105:in `block in logical_paths'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/legacy.rb:104:in `each'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/legacy.rb:104:in `logical_paths'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/manifest.rb:140:in `find'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/sprockets/manifest.rb:185:in `compile'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-rails-3.2.0/lib/sprockets/rails/task.rb:68:in `block (3 levels) in define'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-3.7.1/lib/rake/sprocketstask.rb:147:in `with_logger'
/path/to/vendor/bundle/ruby/2.3.0/gems/sprockets-rails-3.2.0/lib/sprockets/rails/task.rb:67:in `block (2 levels) in define'
/path/to/vendor/bundle/ruby/2.3.0/gems/railties-5.0.1/lib/rails/commands/rake_proxy.rb:14:in `block in run_rake_task'
/path/to/vendor/bundle/ruby/2.3.0/gems/railties-5.0.1/lib/rails/commands/rake_proxy.rb:11:in `run_rake_task'
/path/to/vendor/bundle/ruby/2.3.0/gems/railties-5.0.1/lib/rails/commands/commands_tasks.rb:51:in `run_command!'
/path/to/vendor/bundle/ruby/2.3.0/gems/railties-5.0.1/lib/rails/commands.rb:18:in `<top (required)>'
bin/rails:4:in `require'
bin/rails:4:in `<main>'
Tasks: TOP => assets:precompile
(See full trace by running task with --trace)
zlib(finalizer): Zlib::GzipWriter object must be closed explicitly.
zlib(finalizer): the stream was freed prematurely.
```

**Comment by januszm** [comment]
@yasaichi I get it now, you want to precompile templates inside `app/assets/` . I must say I've never done it this way, is it for a client-side JS frontend? Like Angular?

**Comment by yasaichi** [comment]
@januszm 
It is for static pages like error pages. 
[My gem](https://github.com/yasaichi/gakubuchi) uses this feature to manage the kind of templates by Asset Pipeline.
I'd like to make it work with Sprockets 4, but the build is failing due to the described error 😢

**Comment by januszm** [comment]
It'd be good to check if passing that object doesn't increase memory consumption and so on + if I understand correctly all tests pass?

**Comment by yasaichi** [comment]
@januszm 
> It'd be good to check if passing that object doesn't increase memory consumption and so on

OK, I'll investigate the cost of creating `context` object.

> + if I understand correctly all tests pass?

Sorry I didn’t get it 😢  Could you rephrase that, please?

**Comment by januszm** [comment]
Hey, nvm all tests pass, we're good. Please try to see what happens if passing the context .

**Comment by asartalo** [comment]
Just commenting to say that I also have the same use case as @yasaichi and would love to have this merged

**Comment by januszm** [comment]
@asartalo @yasaichi OK let's give it a second look and try. I'll have more time to work on this next week.

File: test/lib/slim-rails_assets_test.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ class DummyApp < ::Rails::Application
 
       asset_path = File.join(dir, 'app', 'assets', 'html', 'test.slim')
       FileUtils.mkdir_p(File.dirname(asset_path))
-      File.write(asset_path, ".test\n  | hi")
+      File.write(asset_path, %(div class=logical_path\n  | hi))
```
Comment (januszm) [inline test/lib/slim-rails_assets_test.rb]: @yasaichi why are we changing the test from one that parses slim `.test` code for div, to explicit div?

Comment (yasaichi) [reply test/lib/slim-rails_assets_test.rb]: @januszm 
I'm sorry for confusing you.
This change is for the following purposes:

* To check whether the asset helpers work or not
* To keep assertions simple

To accomplish the aboves, I replaced `.test` with `div class=logical_path` (as you may know, the method returns  `"test"`).

If you think it's difficult to understand, I'll make it easier by the other way, for example adding another assertion.

Comment (januszm) [reply test/lib/slim-rails_assets_test.rb]: It would be best if we keep the original test and add new for your test case. Then I guess we could merge it and prepare for relase 3.1.4 or 3.2.0 depending on how big the new feature is.

Comment (yasaichi) [reply test/lib/slim-rails_assets_test.rb]: @januszm 
OK, I'll do that within a few days.

**Comment by yasaichi** [comment]
@januszm
I've added one assertion to check whether asset helpers work in templates.
However, this assertion now meaningless because you didn't make the tests run on Sprockets 4 when you added some changes for Sprockets 4 in #142.
So we should change `Appraisals` to make them run on Sprockets 4, what do you think?

**Comment by hibariya** [comment]
Are there any updates on this? I need this to define helper methods which use `Sprockets::Context`.

**Comment by januszm** [comment]
Let's retest this. We may need to add `app/assets/config/manifest.js` to the Dummy app in tsts

**Comment by januszm** [comment]
@yasaichi  any chance you can rebase on top of current `origin/master` and retest?

**Comment by yasaichi** [comment]
@januszm 
Rebased! Thanks for rebooting my PR after about 4 years!

**Comment by januszm** [comment]
@yasaichi  never too late! 😄

**Comment by januszm** [comment]
There are tests for Sprockets 4 now so this should address previous comment https://github.com/slim-template/slim-rails/pull/135#issuecomment-334987822 (hopefully).  I'm still a bit curious about impact of this on resources consumption but let's move on and release as new feature (minor release). If anyone in this thread has time to benchmark a simple app with and without this change, It'd help before release.

**Review by januszm — APPROVED** [review]
